cmd/satellite/gracefulexit.go

```diff
@@ -98,3 +104,64 @@ func generateGracefulExitCSV(ctx context.Context, completed bool, start time.Tim
 	}
 	return err
 }
+
+func verifyGracefulExitReceipt(ctx context.Context, identity *identity.FullIdentity, nodeID storj.NodeID, receipt string) error {
+	signee := signing.SigneeFromPeerIdentity(identity.PeerIdentity())
+
+	bytes, err := hex.DecodeString(receipt)
+	if err != nil {
+		return errs.Wrap(err)
+	}
+
+	// try to unmarshal as an ExitCompleted first
+	completed := &pb.ExitCompleted{}
+	err = proto.Unmarshal(bytes, completed)
+	if err != nil {
+		// if it is not a ExitCompleted, try ExitFailed
+		failed := &pb.ExitFailed{}
+		err = proto.Unmarshal(bytes, failed)
+		if err != nil {
+			return errs.Wrap(err)
+		}
+
+		err = signing.VerifyExitFailed(ctx, signee, failed)
+		if err != nil {
+			return errs.Wrap(err)
+		}
+		err = checkIDs(identity.PeerIdentity().ID, nodeID, failed.SatelliteId, failed.NodeId)
```

Comment:
Maybe we could verify the nodeID before verify the signature since it's cheaper?

Reply:
valid point. updated